### PR TITLE
Create a design proposal for the raw doc fetcher

### PR DESCRIPTION
**Pull request by mcsaucy**
### TL;DR
1. store raw docs in a (content-addressed, maybe?) service
2. define a manifest that expresses the relationship between those docs
3. sign the manifest with a trustworthy key to anchor our data lineage
4. store the manifest in a database or something (not really the focus of the doc)

### Why are we doing this?
So we have an agreed-upon path for some future work and so it's easier for folks to ramp up and contribute to this. That said, if this is approved by everyone and we change our minds, there's nothing stopping up from throwing this out or updating it.

### Things to look out for
1. does this dovetail well with ongoing work?
2. will this cause any problems now or down the road?
3. will this do what we need it to do?
4. will this scale?

### Meta
I figure a PR is better than a Google Doc for archaeology purposes (so we can look back and see exactly who correctly said "I told you so").

I'm going to throw some reviewers onto this. If you're a reviewer, please suggest changes as needed (or approve if you're content). Once everyone has approved and the discussion has stabilized to the point where everyone seems happy, we can merge this in.

If you aren't a reviewer and happen to see this, please feel free to chime in all the same!

### CODE REVIEW

File: doc/design/record_fetcher_proposal.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+* a unique ID
+* which region and record type the manifest represents (e.g., "USA/FL/Bay/Court")
+* some sort of case identifier
+* the timestamp of collection
+
+A bunch of record-identifying metadata is cool, but we also need to be able to represent the documents that compose the
+record. To that end, we'll require a list of document references. Each reference will contain:
+* a content type (we can just use [MIME
+types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types) here)
+* a URI for where that document's contents are durably stored
```
Comment (GitGerby) [inline doc/design/record_fetcher_proposal.md:49]: Is this to be an external URI or the location within the project where the data resides?

What happens when a document is removed from the public record or goes offline?

Suggest that this be the URI from which the document was fetched along with a timestamp of when the document was fetched.

File: doc/design/record_fetcher_proposal.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+* a digital signature of the rest of the manifest
+	* this is required to anchor our data lineage
+	* key management is out of scope for this document (but the appendix touches on a half-baked approach)
+* support for extension in the future (we can't be locked in to just these fields)
+
+### Storage Endpoints
+We need to be able to store the raw documents and manifests somewhere durable. This document doesn't have strong
+opinions regarding how things are stored, but we need a specific API to ensure safe collection. Our collection API:
+* must allow writing new documents and manifests
+* must not allow for the fetching or altering or existing documents or manifests
```
Comment (GitGerby) [inline doc/design/record_fetcher_proposal.md:67]: s/altering or/altering of/

File: doc/design/record_fetcher_proposal.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+* must enforce authentication to prevent dumping unrelated documents
```
Comment (GitGerby) [inline doc/design/record_fetcher_proposal.md:68]: What is considered an unrelated document?

File: doc/design/record_fetcher_proposal.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    }
+  ]
+}	
+```
+
+Determining the PDAP-controlled URIs of the documents will require uploading them first, and then constructing the
+manifest.
+
+### Garbage Collection
+Since this design requires upload documents before writing out a manifest, it's possible that we'll upload docs and then
```
Comment (GitGerby) [inline doc/design/record_fetcher_proposal.md:125]: requires documents to be uploaded before...

File: doc/design/record_fetcher_proposal.md [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+fail (or neglect) to upload a manifest. This results in orphaned documents, which are just wasted storage. 
+
+To mitigate this, a periodic (daily? weekly?) GC job will run in order to clean up orphaned documents that have existed
```
Comment (GitGerby) [inline doc/design/record_fetcher_proposal.md:128]: This is going to grow to be a rather heavy task as the repository grows; can the service receiving the docs buffer them in such a way that a missing manifest will result in the pending docs being purged before being written to the data store? As described garbage collection resource consumption will grow at best linearly as the manifest and document count increase; even when there's nothing to clean up.

File: doc/design/record_fetcher_proposal.md [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+```python
+with Record(record_type="USA/FL/Bay/Court", case_id="1337") as r:
+  r.AddHtml("label1", html_file1)
+  r.AddHtml("label2", html_file2)
+  r.AddPdf("label3", pdf_file)
+  r.commit()  # upload docs, make and upload manifest
+```
+
+## Appendix
+### A half-baked proposal for JWT-based provenance verification
```
Comment (GitGerby) [inline doc/design/record_fetcher_proposal.md:145]: There's an additional benefit here that's not discussed where records or keys found to be compromised can be easily cleaned up by revoking the key that signed the manifest

File: doc/design/record_fetcher_proposal.md (hunk 1 above)
Comment (bamsammich) [reply doc/design/record_fetcher_proposal.md:49]: +1

The design mentions storing documents on PDAP side so, if the URL is external, is it only intended to record origin? Is that valuable if/when the original document is no longer available?

Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md]: Sorry, I meant URI where this data can be fetched internal to the project, not where we grabbed it from the internet. I don't think we need a timestamp here if we include it in the manifest (having per-document granularity provides limited utility, IMO), but I'd be interested in hearing arguments for it.

File: doc/design/record_fetcher_proposal.md (hunk 3 above)
Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md]: Expanded upon this a bit. I was mostly interested in untrusted entities soaking up all of our money by trying to store /dev/urandom of whatever.

File: doc/design/record_fetcher_proposal.md (hunk 5 above)
Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md]: Added a different approach that requires a bit more coordination, but doesn't require us to enumerate potentially millions of documents (assuming this thing takes off). PTAL.

File: doc/design/record_fetcher_proposal.md (hunk 6 above)
Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md:153]: Agreed. Expanded upon that a bit at the bottom.

**Comment by nfrostdev** [comment]
This is great! I'm trying to get more eyes on it to see if anyone has anything to nitpick.

File: doc/design/record_fetcher_proposal.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+## BLUF
+
+Upload fetched documents, having performed the minimal amount of processing. Define a manifest expressing the
+association between those document URIs and basic metadata (ID, region and record type, timestamp of collection, etc.).
+Sign that manifest and upload it. Now a logical record's components may be derived from the manifest and the now-stored
+documents the manifest describes.
+
+## Objective
+
+Define a system for tracking the associations between the various crude scrapings that describe a record. How documents
```
Comment (tthelin) [inline doc/design/record_fetcher_proposal.md:12]: doc nit: Can you split the second sentence into a new paragraph, it reads a little better

File: doc/design/record_fetcher_proposal.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+* the timestamp of collection
+
+A bunch of record-identifying metadata is cool, but we also need to be able to represent the documents that compose the
+record. To that end, we'll require a list of document references. Each reference will contain:
+* a content type (we can just use [MIME
+types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types) here)
+* a digest of the content
+* a URI for where that document's contents are durably stored in PDAP infrastructure
+  * this isn't necessary if we adopt storage infrastructure that allows us to reliably derive this from the digest
+* an optional comment expressing where this specific document came from
```
Comment (tthelin) [inline doc/design/record_fetcher_proposal.md:52]: If we want to achieve reproducability (automate pulling the data again for any reason), this seems required per manifest

File: doc/design/record_fetcher_proposal.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+need to be able to track distinct snapshots over time. Additionally, we need a service which will take this information
+and persist it in a meaningful way. Analysis of this information is not in scope.
+
+### Manifests
+
+A manifest needs to be able to represent the raw materials associated with a record. The minimal identifiers for the
+record itself are:
+* a unique ID
+* which region and record type the manifest represents (e.g., "USA/FL/Bay/Court")
+* some sort of case identifier
```
Comment (tthelin) [inline doc/design/record_fetcher_proposal.md:42]: I'd personally view the case identifier as part of the records "fields" that will be extracted later based on the type of the record.

If we store different kinds of records, case identifier isn't universal.  Unique ID, timestamps, origin, PDAP URI, etc are all universal across any kind of record type

File: doc/design/record_fetcher_proposal.md [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+We need a way to store raw information pertaining to logical records. Given logical records can span documents, we also
+need to store a representation (a manifest) of those relationships. Given records are capable of changing over time, we
+need to be able to track distinct snapshots over time. Additionally, we need a service which will take this information
+and persist it in a meaningful way. Analysis of this information is not in scope.
+
+### Manifests
+
+A manifest needs to be able to represent the raw materials associated with a record. The minimal identifiers for the
+record itself are:
+* a unique ID
```
Comment (tthelin) [inline doc/design/record_fetcher_proposal.md:40]: globally unique? specific format?

If there's not another more concrete proposal I'd vote for UUIDv4 as they are widely supported everywhere making them easy to use
https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)

File: doc/design/record_fetcher_proposal.md [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+* some sort of case identifier
+* the timestamp of collection
+
+A bunch of record-identifying metadata is cool, but we also need to be able to represent the documents that compose the
+record. To that end, we'll require a list of document references. Each reference will contain:
+* a content type (we can just use [MIME
+types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types) here)
+* a digest of the content
+* a URI for where that document's contents are durably stored in PDAP infrastructure
+  * this isn't necessary if we adopt storage infrastructure that allows us to reliably derive this from the digest
```
Comment (tthelin) [inline doc/design/record_fetcher_proposal.md:51]: I'd be careful of doing a direct tie of a PDAP URI to how we're storing it, as that might change over time.  It seems like we should be able to come up with an abstract yet simple PDAP URI concept  pdap://<conceptual storage phase>/uuid (or similar)

File: doc/design/record_fetcher_proposal.md [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+* an optional comment expressing where this specific document came from
+
+In order for us to trust our collected record results, we need to be able to verify that we trust the software which
```
Comment (tthelin) [inline doc/design/record_fetcher_proposal.md:54]: Not disagreeing, but I hope we'll be able to come up with some per-language libraries to help with this, it could be daunting for the average person

File: doc/design/record_fetcher_proposal.md [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+  * this is required to anchor our data lineage
+  * key management is out of scope for this document (but the appendix touches on a half-baked approach)
+* support for extension in the future (we can't be locked in to just these fields)
+
+### Storage Endpoints
+We need to be able to store the raw documents and manifests somewhere durable. This document doesn't have strong
+opinions regarding how things are stored, but we need a specific API to ensure safe collection. Our collection API:
+* must allow writing new documents and manifests
+* must not allow for the fetching or altering of existing documents or manifests
+* must enforce authentication and authorization for document submission
```
Comment (tthelin) [inline doc/design/record_fetcher_proposal.md:70]: I was having this conversation somewhere else, but to me there's a lot of value in not allowing outside data submissions for things that can be fetched.  People submit code, and then we run it it in an appropriate environment where we can control the chain of trust ourselves, allowing for a simpler execution/trust model.

However even within a trusted environment, the entire manifest concept is still a good idea because we need a solid audit trail for all of the data

File: doc/design/record_fetcher_proposal.md [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+  * this lets us prevent unknown/untrusted entities from aiming a firehose of data at us and eating all of our money,
+  but there's little risk of this threatening the trustworthiness of our records due to the lack of manifest
+  * this document doesn't have a strong opinion here regarding how we authenticate, but the appendix has a strawman we
+can use to start that conversation
```
Comment (tthelin) [inline doc/design/record_fetcher_proposal.md:74]: doc nit: bullet formatting alignment in this section

File: doc/design/record_fetcher_proposal.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+  "record_type": "USA/FL/Bay/Court",
+  "case_id": "20000001",
+  "collected_at_ms": 1592618026000,
+  "fetcher_id": "github.com/Police-Data-Accessibility-Project/Scrapers:USA/FL/Bay/Court/scraper/Scraper.py@1db20ce1e63768363d532d46a641722986f63524",
+  "documents": [
+    {
+      "label": "some pdf we'll have to ocr",
+      "content_type": "application/pdf",
+      "digest": "...",
+      "uri": "https://..."
```
Comment (tthelin) [inline doc/design/record_fetcher_proposal.md:114]: This makes it seem like every document is stored individually?  I was assuming the "record" was a bundle that's atomically stored as a unit. It would provide easier access patterns (you fetch and save everything together) for instance.  Either all of it would be stored or none of it, avoiding issues of partially uploaded data.

It should be easy to manage bundles using MIME Multipart, which is widely supported in various languages too
https://en.wikipedia.org/wiki/MIME#Multipart_messages

Is there an advantage to keeping everything separate?

Comment (GitGerby) [reply doc/design/record_fetcher_proposal.md:114]: I assumed a "record" was a specific event or case that my span multiple retrieved documents, storing the documents with an audit trail and the auditably constructing the "record" from source would give a chain of custody and allow anyone to reproduce the record from the source material.

File: doc/design/record_fetcher_proposal.md (hunk 9 above)
Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md:43]: Could you expand upon this some?

If I'm understanding you correctly, we'd analyze the record (post-fetching, of course) and then derive some sort of record identifier based upon that?

File: doc/design/record_fetcher_proposal.md (hunk 11 above)
Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md]: I'd been trying to not say "all stored documents are content-addressed", but I'm tired of dancing around that. Since we don't really have anything else to go off of, I'm gonna require that.

File: doc/design/record_fetcher_proposal.md (hunk 8 above)
Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md]: I've added the label requirement (which I'd accidentally omitted previously). I think that should be sufficient, unless I'm misunderstanding.

File: doc/design/record_fetcher_proposal.md (hunk 12 above)
Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md:54]: Oh, I totally agree and that is the intention.

File: doc/design/record_fetcher_proposal.md (hunk 13 above)
Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md:70]: Strongly agreed.

File: doc/design/record_fetcher_proposal.md (hunk 15 above)
Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md]: > This makes it seem like every document is stored individually?

That was the intent.

>  I was assuming the "record" was a bundle that's atomically stored as a unit.

That is a *way* better idea! Completely removes the garbage collection requirement by preventing the scenario from ever arising.

File: doc/design/record_fetcher_proposal.md (hunk 9 above)
Comment (nfrostdev) [reply doc/design/record_fetcher_proposal.md:43]: I think what he was saying is that an attribute like this may be difficult at the moment to proclaim that it could be at a high level like this in the manifest?

His suggestion may have been slightly off base for what you were intending the text to contain below line 39/40.

File: doc/design/record_fetcher_proposal.md [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+
+```json
+{
+  "version": 1,
+  "uuid": "6e566397-6803-49ec-9280-62c00b368e3f",
+  "record_type": "USA/FL/Bay/Court",
+  "case_id": "20000001",
+  "collected_at_ms": 1592618026000,
+  "fetcher_id": "github.com/Police-Data-Accessibility-Project/Scrapers:USA/FL/Bay/Court/scraper/Scraper.py@1db20ce1e63768363d532d46a641722986f63524",
+  "documents": [
```
Comment (nathanmentley) [inline doc/design/record_fetcher_proposal.md:105]: Hmm. I'm kind of confused, and feel free to let me know if I'm missing something or over complicating things.

It feels like we'd be adding data related to the request in the JWT, If we're adding a signature related to the specific request shouldn't it be in another header, and not connected to the signed jwt?
Since, I think this is essentially a checksum on the data, It's solving a separate problem from auth/auth and and probably should just be included elsewhere in the request.
Maybe we could store each part's checksum in a PDAP_PART_N_DIGEST header or something similar to that? Or a complex object of signatures in a PDAP_DIGEST header.

File: doc/design/record_fetcher_proposal.md [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+## Caveats
+This is somewhat complex and may be tricky to implement. A way to mitigate this is to invest in easy-to-use canonical
+implementations in common languages with an API that makes misuse hard. For example, a Python API could resemble:
+
+```python
+with Record(record_type="USA/FL/Bay/Court", case_id="1337") as r:
+  r.AddHtml("label1", html_file1)
+  r.AddHtml("label2", html_file2)
+  r.AddPdf("label3", pdf_file)
+  r.commit()  # construct a manifest, upload bundle of manifest + documents
```
Comment (nathanmentley) [inline doc/design/record_fetcher_proposal.md:141]: I like how fluent and readable this is.

Typically in a builder pattern you'd expect the builder to return itself. So you could do:

r
  .AddHtml("label1", html_file1)
  .AddHtml("label2", html_file2)
  .AddPdf("label3", pdf_file)
  .commit()

I'll be honest, I'm no python expert. So maybe that's not super pythonic, but regardless of the builder returning it self or just calling on "r" each time, this is super readable and I just thought it was worth while to point that out.

Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md:141]: Java really lends itself to that kind of expressions, but we'd have to keep Python from ending the expression prematurely. Something like:
```python
r\
.addHtml("label1", html_file1)\
.addHtml("label2", html_file2)\
.commit()
```
Which is kinda gross, or:
```python
(r
).addHtml("foo", bar
).commit()
```
... which is absolutely cursed.

Anyway, thanks for the feedback! I'm a big fan of try to make hard things simple. 😄

File: doc/design/record_fetcher_proposal.md (hunk 9 above)
Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md:43]: Hmmm. I think I understand this better now. Thanks for the help!

So I don't think I care about the actual case number for my imaginary concerns, it's more that I want any number that can be used to link two collected records together. If we fetch the same content twice and the document digests don't match up exactly, it'll take a lot of processing to recognize those are actually the same if we don't have some stable identifier. That said, I can totally see this being hard not just for online records, but if we try to apply the same structure to scanned documents or whatever.

In some situations, the case number is well established. For example, the Benchmark portal is case number-centric, so we *know* the number associated for those cases because we had to explicitly search for them. In those cases, we have a strong confidence in that identifier because we asked explicitly for that number and that's what the source gave us.

How would folks feel about making this optional? Or do we still think we should just not collect it? If we drop it now, we can always change out minds later (it may just require re-fetching). Thoughts?

File: doc/design/record_fetcher_proposal.md [hunk 18]
```diff
@@ -37,19 +38,18 @@ and persist it in a meaningful way. Analysis of this information is not in scope
...
 types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types) here)
-* a digest of the content
-* a URI for where that document's contents are durably stored in PDAP infrastructure
-  * this isn't necessary if we adopt storage infrastructure that allows us to reliably derive this from the digest
-* an optional comment expressing where this specific document came from
+* a digest of the content, which may be used to later fetch the document from PDAP-owned storage
+* an optional comment, potentially expressing where this specific document came from
 
 In order for us to trust our collected record results, we need to be able to verify that we trust the software which
 performed the collection. This means being able to audit the code that created each manifest. To this end we also need
```
Comment (jameskranz) [inline doc/design/record_fetcher_proposal.md:55]: Just for clarity, is the plan not to run all scrapers in the cloud under our control and to also generate the docker images we use directly from the source in our own build process?

I've had to be away from this for a bit, so I've without a doubt missed details as they developed.

File: doc/design/record_fetcher_proposal.md [hunk 19]
```diff
@@ -77,22 +77,18 @@ can use to start that conversation
...
-structure for this, as any relationship between documents will be encoded by the manifests -- the service just needs to
-be able to store the raw documents for later lookup. Backing this storage API with a fairly flat filesystem is fine
-here, we're rarely going to list files, only write and read. If we do need to list files, it'll be for some cleanup
-operation and should not be on any request path.
+First, let's talk document storage. This will be a content digest-addressed blob storage service. We don't need a
+lot of structure for this, as any relationship between documents will be encoded by the manifests -- the service just
+needs to be able to store the raw documents for later lookup (but lookups don't need to be via this service). Backing
+this storage service with a fairly flat filesystem is fine here, we're rarely going to list files, only write and read.
 
 A big benefit of content-addressed storage here is that we get deduping almost for free -- we may need to zero out
```
Comment (jameskranz) [inline doc/design/record_fetcher_proposal.md:87]: I'd posit that we'd want to keep a record of material changes, which this makes difficult. Raw docs stored for this would be expensive from a storage standpoint, but diffs could be an interesting alternative.

File: doc/design/record_fetcher_proposal.md [hunk 20]
```diff
@@ -111,31 +107,27 @@ manifest's information in the JWT payload. As an example, the payload could rese
...
-manifest.
-
-### Garbage Collection
+### Record Bundles
 
-Since this design requires documents to be uploaded before writing out a manifest, it's possible that we'll upload docs
-and then fail (or neglect) to upload a manifest. This results in orphaned documents, which are just wasted storage. 
-To mitigate this, we'll construct a GC batch job that runs on a schedule and removes documents that are orphaned and
-have existed for longer than X time.
+Signed manifests and all related documents will be uploaded together as [MIME Multipart messages](
```
Comment (jameskranz) [inline doc/design/record_fetcher_proposal.md:121]: Just reiterating that I'm not sure this is a concern if this all happens under our control.

File: doc/design/record_fetcher_proposal.md (hunk 16 above)
Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md:105]: If I'm understanding your concerns correctly, I touch on this in the appendix, but I didn't want to make its suggestions a hard requirement or anything because it's A Lot. The motivation here is that I want every record we fetch to be associable with the code that is running and how it got there, and I want that association to be tamper proof (or at least glaringly tamper evident) because if this takes off, I could see parties interested in tampering with records.

I'm kinda envisioning two JWTs. One that is a bearer token used to authenticate the fetcher, and one that is really a signed encapsulation of our manifest (where only the fetcher holds the private key). The signed manifest format doesn't *have* to be JWT, but it meets the requirements pretty well: signs a JSON payload, flexible crypto options, seems to have good library support in many languages.

As for the digests, while they are checksums, if we use cryptographically strong checksums (say sha256) and persist them to our persistent storage, we can detect tampering of the raw documents. Similarly, by signing all manifests and persisting those signatures to our persistent storage, we can detect tampering at any point in the future by verifying the signature. So by verifying the signature, we can detect tampering in the manifest, which allows us to detect tampering of the underlying documents.

Sorry if this misses what you were going for. Does that help? If so, how could I make this more clear in the document?

File: doc/design/record_fetcher_proposal.md (hunk 18 above)
Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md:55]: That is still the plan. This design doesn't work in an environment where we cannot trust the infrastructure.

Something that may have changed is that we're moving to a model where we split raw record collection (in the form of multiple documents) and the extraction of information from those documents. This document strives to:
1. define a mechanism for representing the relationships between those raw documents
2. make any tampering evident
3. leave us something we can latch on to in order to verify that all data came from PDAP code and infra

Comment (jameskranz) [reply doc/design/record_fetcher_proposal.md:55]: Can't 3 be ensured through some combination of mtls and by not having the message queue / upload interface be available beyond the cloud network? What would the 'story' of tampering look like if everything was internal to our network, other than the scrape target? (aka the court record website)

File: doc/design/record_fetcher_proposal.md (hunk 20 above)
Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md:121]: mTLS was the original plan, but that just gets us endpoint auth, so I ditched that in favor of the JWT bearer token (not to be confused with the signed manifest JWT) because it's simpler. That said, I wanted things to be tamper evident even while in PDAP-controlled storage, as I don't really trust PDAP-controlled storage to not be tampered with.

We're working on exposing government accountability data to journalists, activists, lawmakers, etc. in situations where it's difficult to collect records to begin with (and therefore difficult to verify our data), so tampering with our data could be highly valuable for some parties. We're a loose coalition of volunteers where no one knows anyone (some folks aren't even using their real names), so we can't *really* put a ton of trust in anyone, and I don't know who would have write access to the raw data.

Targeted tampering with records seems low effort and high reward, so I'm trying to make that tampering evident at a fundamental level. Thoughts?

Comment (jameskranz) [reply doc/design/record_fetcher_proposal.md:121]: The idea of not trusting PDAP-controlled storage is a bit concerning to me. Hopefully, you mean you don't trust what is stored there, which is more reasonable. The question of where to draw lines of what is trustworthy is still there, though. If we can't trust what is running within our infrastructure, how can we trust the infrastructure managing JWT tokens and their authentication?

**Comment by danmelles** [comment]
Overall thought: I think that creating a cataloguing system is the most important element here. This is because much of the challenge we're trying to address is organization of data, not as much fetching it. In theory, people could grab data manually and upload it to a data lake.
The reason I bring this up is that I think it's an important enough issue to warrant its own design doc, separate from anything related to scraping. Thoughts?

File: doc/design/record_fetcher_proposal.md (hunk 20 above)
Comment (mcsaucy) [reply doc/design/record_fetcher_proposal.md:121]: > The idea of not trusting PDAP-controlled storage is a bit concerning to me. Hopefully, you mean you don't trust what is stored there, which is more reasonable.

Yeah, sorry about that. "Not trusting storage" has implications I didn't mean to convey. From an insider risk angle (or just misconfiguration) angle, I'm hesitant to assume that nothing in our storage has been tampered with. I think what I'm mostly shooting for is a model where we largely trust our storage and its contents, but are able to scalably verify they are trustworthy.

> EDIT: Just as an additional note, security is not really my wheelhouse.

My present work is security adjacent, but this is still largely not my wheelhouse either. 😄 I may pop a link to this in the infosec chat, since the discussions surrounding this has drifted into that territory. On the topic of things that aren't wheelhouses, I'm not well-versed in what the actual landscape looks like here (e.g., is there something in AWS that makes some aspect of this moot?), so there could be some assumptions I'm incorrectly making/not-making.

> If we can't trust what is running within our infrastructure, how can we trust the infrastructure managing JWT tokens and their authentication?

We have to root our trust somewhere, but we can confine it to a something that is extremely locked down and audited (such as the key infra). The strawman in the appendix addresses some of this, but let's say:
1. fetcher generates a keypair (private key stays in memory so it's lost on process exit)
2. fetcher registers the public key and provenance info (repo, entrypoint, commit, etc) with infra, which logs that data in a DB (we'll call this abstract thing the _keydb_) and returns a JWT bearer token for auth
3. manifests are signed with generated key
4. upload endpoint ensures JWT token is legit, validates the manifest's signature with the public key in the JWT bearer token

> I'd love to learn more and understand how this makes it safer, or if the goal is more being able to track exactly where data came from and what processed it.

I'm shooting for both.

When it comes to safety, the following scenarios can play out if we get owned:
* manifest/doc storage compromised:
  * action: attacker manages to get write access our manfiest/document store and changes a bunch of things
  * response: we'd verify the signatures of all manifests against what we have in the keydb and find anything that was inserted or tampered with (no solution for deletions)
* fetcher environment compromised:
  * action: attacker can run code in the fetcher environment to upload bad record data
  * response: we can identify the key used to upload the bad data and revoke the key in the keydb, causing verifications to fail
* fetcher's key material leaks:
  * action: a fetcher's private key is exfil'd and it's no longer actually private
  * response: revoke the key in the keydb, causing verifications to fail
* keydb is compromised
  * action: attacker can write arbitrarily to the keydb
  * response: if you can identify the tampering, revoke the relevant entries. If not, blow it away, restore from more trusted backups, re-fetch what you need to re-fetch.

As for tracking origin, this would allow us to confidently say "this data came from this scraper at this revision and this time", and then other things can build off of that (with processes which are ideally also verifiable).

Thoughts?

**Comment by GitGerby** [comment]
Asserting the authenticity of the data is the challenge here, processing
and cataloging can be done as needed to present the underlying records in a
useful manner. Ensuring that the data is correct and free from manipulation
is paramount, compromised source materially would fundamentally undermine
the entire project.

File: doc/design/record_fetcher_proposal.md (hunk 18 above)
Comment (nfrostdev) [reply doc/design/record_fetcher_proposal.md:55]: I'm not really sure how to respond to James here other than really just mitigation of potential source overload.

If we were to blame for a denial of service attack because someone was very aggressively scraping data and funneling it to us it could look bad.

Comment (jameskranz) [reply doc/design/record_fetcher_proposal.md:55]: My disconnect may just come from my (erroneous?) assumption that people will contribute scrapers, but they will be executed from within our control after being audited (pr review). No bad actor could funnel data (good or bad) to us in that model.

File: doc/design/record_fetcher_proposal.md (hunk 9 above)
Comment (OscarVanL) [reply doc/design/record_fetcher_proposal.md:43]: Some kind of record-specific identifier is a must to avoid duplication as mentioned by Josh. Case ID is the most obvious, but the absence of case IDs is inevitably going to happen on some portal(s). If we drop Case IDs, what takes its place?

In this case, it's important to consider how we should consistently create the same identifier when scraping a given court record multiple times. A crude way would be to take a bunch of fields and hash them to make an ID. Better ideas are welcomed...
